File: solutions/csharp/2141-maximum-running-time-of-n-computers.cs
public class Solution {
    public long MaxRunTime(int n, int[] batteries) {
        Array.Sort(batteries);
        long left = 1, right = (long)batteries.Sum(b => (long)b) / n;
        while (left < right) {
            long target = right - (right - left) / 2;
            long total = batteries.Sum(battery => Math.Min((long)battery, target));
            if (total >= target * n) {
                left = target;
            } else {
                right = target - 1;
            }
        }
        return left;
    }
}

File: solutions/cpp/2141-maximum-running-time-of-n-computers.cpp
class Solution {
public:
    long long maxRunTime(int n, std::vector<int>& batteries) {
        std::sort(batteries.begin(), batteries.end());
        long long left = 1, right = std::accumulate(batteries.begin(), batteries.end(), 0LL) / n;
        while (left < right) {
            long long target = right - (right - left) / 2;
            long long total = std::accumulate(batteries.begin(), batteries.end(), 0LL, [&target](long long a, int b) { return a + std::min(static_cast<long long>(b), target); });
            if (total >= target * n) {
                left = target;
            } else {
                right = target - 1;
            }
        }
        return left;
    }
};
